I2C API Enforce stop condition after last message in `i2c_transfer` wrapper

bjarki-andreasen opened the pull request:
The I2C API specifies that the last message of a transfer implies a STOP condition. This requirement has been offloaded to each device driver to implement, which results in many in-tree drivers (more than half of them) not following this requirement.

Since the STOP is implied by the API, we can ensure the STOP bit of the last message is set before passing the messages to the device drivers for transfer. This ensures all device drivers implicitly follow this requirement, allowing us to remove this now duplicate code from the in-tree device drivers which currently do this internally.

We can additionally prevent calls to the `transfer` and `transfer_cb` APIs if `num_msgs` is 0, which is a check we need to do anyway before setting the last messages STOP bit, allowing us to return early if there is nothing to transmit, as is described by the API as well: "@a num_msgs may be zero,in which case no transfer occurs."

bjarki-andreasen-nordic committed:
drivers: i2c: enforce i2c stop flag in last i2c_transfer msg

The i2c_transfer API specifies that the I2C_MSG_STOP flag implicitly
is set for the last message in a i2c_transfer(). This is identical to
explicitly setting the I2C_MSG_STOP flag in the last message, which is
what every driver which properly adheres to this specification does.

This commit updates the i2c_transfer() API wrapper to explicitly set
the flag before calling the drivers API implementation, which ensures
the specification is followed, while moving this additional complexity
from the device drivers themselves.

Additinally it checks for the num_msgs being 0, in which case no
transfer shall occur, nor shall any stop condition be set. This removes
an additional check from device drivers.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: i2c: i2c_dw: Remove redundant last msg stop

The I2C API itself now enforces last msg stop, remove duplicate code
from driver.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: i2c: i2c_esp32: Remove redundant last msg stop

The I2C API itself now enforces last msg stop, remove duplicate code
from driver.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: i2c: i2c_gd32: Remove redundant last msg stop

The I2C API itself now enforces last msg stop, remove duplicate code
from driver.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: i2c: i2c_ll_stm32: Remove redundant last msg stop

The I2C API itself now enforces last msg stop, remove duplicate code
from driver.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

gmarull commented on include/zephyr/drivers/i2c.h:
```diff
@@ -531,6 +531,15 @@ static inline void i2c_dump_msgs(const struct device *dev, const struct i2c_msg
 	i2c_dump_msgs_rw(dev, msgs, num_msgs, addr, false);
 }
 
+/** @cond INTERNAL_HIDDEN */
+
+static inline void i2c_set_last_msg_stop(struct i2c_msg *msgs, uint8_t num_msgs)
+{
+	msgs[num_msgs - 1].flags |= I2C_MSG_STOP;
+}
+
+/** @endcond */
```
is this inline helper really necessary?

bjarki-andreasen replied:
no, but it is pretty :)

ydamigos replied:
`msgs[num_msgs - 1].flags |= I2C_MSG_STOP;` is straight forward.
I don't see the point of adding 8 lines of code if there is only one call to the inline function.

bjarki-andreasen replied:
Understood :)

bjarki-andreasen commented:
Removed the wrapper function :)

teburd commented on include/zephyr/drivers/i2c.h:
```diff
@@ -785,6 +785,8 @@ static inline int z_impl_i2c_transfer(const struct device *dev,
 	const struct i2c_driver_api *api =
 		(const struct i2c_driver_api *)dev->api;
 
+	msgs[num_msgs - 1].flags |= I2C_MSG_STOP;
```
Please add an assert before this verifying num_msgs >= 1

bjarki-andreasen replied:
ok, did not notice we allow a length of zero :) Out of curiosity, why do we allow this?

bjarki-andreasen-nordic committed:
drivers: i2c: enforce i2c stop flag in last i2c_transfer_cb msg

The i2c_transfer API specifies that the I2C_MSG_STOP flag implicitly
is set for the last message in a i2c_transfer(). This is identical to
explicitly setting the I2C_MSG_STOP flag in the last message, which is
what every driver which properly adheres to this specification does.

The i2c_transfer_cb API is seemingly identical to the i2c_transfer_api,
it references it using @see i2c_transfer(), while providing a callback,
and being available from isr context. This commit extends the wrapper
to provide the following guarantees also present in the i2c_transfer API:

- If num_msgs is 0, no transfer will occur
- The last message of the transfer implies a stop condition

This allows for users to use both functions interchangeably, while
removing handling of num_msgs = 0 and last message missing stop cond
from device drivers.

Additionally a check for num_msgs == 0 is added to prevent any transfer
and immediately call back with success then return without invoking
the transfer_cb API.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
I have updated the wrapper to check if `num_msgs == 0` and immediately return instead of invoking the `transfer` implementation given that no transaction will occur, so there is no reason to call it. This further optimized device drivers since they don't have to do this check internally.

Additionally, I implemented the set stop condition + check for `num_msgs == 0` to the async version of `i2c_transfer`, `i2c_transfer_cb` which I did not notice initially :)
